sigmastate/src/main/scala/sigmastate/eval/Profiler.scala

```diff
@@ -0,0 +1,325 @@
+
+  /** Update time measurement stats for a given operation. */
+  final def addPoint(key: K, point: V) = {
+    val item = opStat.getOrElse(key, null)
+    if (item != null) {
+      item.addPoint(point)
+    } else {
+      val item = new StatItemImpl
+      item.addPoint(point)
+      opStat(key) = item
+    }
+  }
+
+  final def mapToArray[@sp C: ClassTag](f: (K, StatItem[V]) => C): Array[C] = {
+    opStat.mapToArray(f)
+  }
+}
+
+/** A simple profiler to measure average execution times of ErgoTree operations. */
+class Profiler {
+
+  // NOTE: this class is mutable so better to keep it private
+  private class OpStat(
+    /** The node on the evaluation stack. */
+    val node: SValue,
+    /** The time then this node evaluation was started. */
+    val outerStart: Long,
+    /** The accumulated time of evaluating all the sub-nodes */
+    var innerTime: Long,
+    /** The time then this nodes evaluation finished */
+    val outerEnd: Long
+  )
+
+  /** If every recursive evaluation of every Value is marked with
+    * [[onBeforeNode()]]/[[onAfterNode()]], then this stack corresponds to the stack of
+    * recursive invocations of the evaluator. */
+  private var opStack: List[OpStat] = Nil
+
+  /** Called from evaluator (like [[sigmastate.interpreter.ErgoTreeEvaluator]])
+    * immediately before the evaluator start recursive evaluation of the given node.
+    */
+  def onBeforeNode(node: SValue) = {
+    val t = System.nanoTime()
+    opStack = new OpStat(node, t, 0, t) :: opStack
+  }
+
+  /** Called from evaluator (like [[sigmastate.interpreter.ErgoTreeEvaluator]])
+    * immediately after the evaluator finishes recursive evaluation of the given node.
+    */
+  def onAfterNode(node: SValue) = {
+    val t = System.nanoTime()
+
+    val op = opStack.head   // always non empty at this point
+    opStack = opStack.tail  // pop current op
+    assert(op.node.opCode == node.opCode, s"Inconsistent stack at ${op :: opStack}")
+
+    val opFullTime = t - op.outerStart  // full time spent in this op
+
+    // add this time to parent's innerTime (if any parent)
+    if (opStack.nonEmpty) {
+      val parent = opStack.head
+      parent.innerTime += opFullTime
+    } else {
+      // we are on top level, do nothing
+    }
+
+    val opSelfTime = opFullTime - op.innerTime
+
+    // update timing stats
+    addOpTime(node.opCode, opSelfTime)
+  }
+
+  /** Timings of op codes. For performance debox implementation of Map is used. */
+  private val opStat = new StatCollection[Int, Long]()
+
+  /** Update time measurement stats for a given operation. */
+  @inline private final def addOpTime(op: OpCode, time: Long) = {
+    opStat.addPoint(OpCode.raw(op), time)
+  }
+
+  /** Timings of method calls */
+  private val mcStat = new StatCollection[Int, Long]()
+
+  /** Update time measurement stats for a given method. */
+  @inline private final def addMcTime(typeId: Byte, methodId: Byte, time: Long) = {
```

Comment:
method not used

Reply:
I used it before and it may be useful later, since Profiler is utility class, I would leave it there for a while.